src/vltrace.c

```diff
@@ -261,42 +250,27 @@ try_to_read_lock_file()
 static int
 check_if_only_instance(void)
 {
-	int pid = 0;
-	FILE *sfd;
+	char *buf = NULL;
 
-	int fd = open(FILE_LOCK, O_WRONLY | O_CREAT | O_EXCL, 0644);
-	if (fd == -1) {
-		/*
-		 * Could not create a new lock file,
-		 * so try to open and read it.
-		 */
-		pid = try_to_read_lock_file();
-		if (pid != -1)
-			/* vltrace is already running */
-			return pid;
-
-		/* reading lock file failed, so try to write the current pid */
-		unlink(FILE_LOCK);
-		WARNING(
-			"starting vltrace, however another instance may be running");
-		sfd = fopen(FILE_LOCK, "w");
-	} else {
-		sfd = fdopen(fd, "w");
+	FD_lock = open(FILE_LOCK, O_RDWR | O_CREAT, 0644);
+	if (FD_lock == -1) {
+		perror(FILE_LOCK);
+		return -1;
 	}
 
-	if (sfd != NULL) {
-		/* write PID to the lock file */
-		pid = getpid();
-		if (fprintf(sfd, "%i \n", pid) < 0)
-			WARNING("error writing to the lock file");
-		fclose(sfd);
-	} else {
-		perror("fopen");
-		WARNING("error opening lock file for writing");
+	if (flock(FD_lock, LOCK_EX | LOCK_NB)) {
+		/* cannot grab the file lock, try to read the PID from file */
+		return try_to_read_lock_file();
+	}
+
+	/* write PID to the lock file */
+	(void) asprintf(&buf, "%i \n", getpid());
```

Comment:
I suspect I know why you put space before \n, but this is ugly approach - just truncate the file before writing anything.